### DISCUSSION TITLE
ZeroTier networks

### DISCUSSION THREAD

**Discussion by CommonLoon102** [diasurgical/DevilutionX#6471] [Q&A] [answered]
Hello @AJenbo (or whoever can answer),

I'm interested in the network "strategy" you have chosen for ZeroTier. Every player are on the same network or when you create a new lobby that would create a new network in ZT?

Are there any performance impact having every player in one network instead of having a network for each lobby? What I could think of that ZT is selecting different relays if a network has different peers. If one network has all the peers (players), even those who don't play together, then ZT might select sub-optimal relays (if relays are needed). If relays are selected on a per-connection basis, then this is a non-issue.

How have you implemented password protected games? If all players are on the same ZT network, then I guess you have added a separate layer of encryption over it. In one of the issues libsodium is mentioned, is this its purpose? If each lobby would be on a separate network then this additional encryption wouldn't be needed, would it?

**Comment by AJenbo** [comment]
There normally aren't more then about a dozen players at any time so it's not a big concern.

Yes, private games are encrypted using libsodium, this isn't related to the network protocol being used and was there before we added ZT support.

I'm not sure what you are referring to as a "lobby", if each player is on their own network they won't be able to discover other games or join any other game, and what other information would you use to partition the players? Would they have to manually enter a shared secret network ID? That would break the idea of public discoverable games.

Since ZT isn't the only supported protocol and it's not directly tied to encryption, changes to how ZT is managed won't make libsodium encryption redundant.

P.s. I'm not the only person on this project and I didn't implement the network and I don't do the majority of the maintenance on it either probably better to state the question in general than directed at me.

P.s.s. I see you are active on NBlood, cool :+1:

**Reply by CommonLoon102** [reply]
The lobby is a game. Two paladins and a barbarian play together. There might be other players but they are in different games (lobbies).

Lobbies (and their corresponding ZT network IDs) could be listed by a master server. When a player selects which game (lobby) they want to join to, the program would join to the corresponding ZT network ID. I thought it is working like this, but it seems it doesn't, and I wanted to know why.

**Reply by qndel** [reply]
because nobody is going to manage countless ZT networks and it works fine the way it does at the moment?

**Reply by AJenbo** [reply]
Then we would need to create a master server and implement a protocol for that. This would also then need to manage game discovery as we would not be able to have that happen via ZT any longer.

That would be a lot of work and maintenance, plus we would have expenses for running the master server.

We don't have any indication that there would be any issues with how ZT manage the connections and preferably connections would not be relayed at all and instead successfully establish direct connections between players.

It would also be a point of failure that could bring down multiplayer which is also preferable to avoid.

**Reply by CommonLoon102** [reply]
I think I understand now. Diablo can detect LAN games on its own I guess. Maybe by broadcasting a packet and then every machine on the same network will respond if it has hosted a game. In NBlood we don't have such discoverability, so we would need to implement something anyways.

If each game is on a separate ZT network, then it is not needed to put encryption onto the already encrypted ZT network. The ZT network ID won't be announced to the master server. This way the password is the ZT network ID. Just make it sure it can be easily copy-pasted.

**Reply by StephenCWills** [reply, marked as answer]
> Lobbies (and their corresponding ZT network IDs) could be listed by a master server.

There is no master server, so that would be a good reason not to do it this way.

> When a player selects which game (lobby) they want to join to, the program would join to the corresponding ZT network ID.

Have you ever used the libzt library? AFAIK, there's no way to spin up a new network on the fly. Maybe a master server could do it with a paid account and admin credentials? Sounds like a pretty complex implementation.

**Reply by CommonLoon102** [reply]
No, I haven't used libzt yet.
If creating a new network is not just a function call to the library, then it is not an option at all.

Thank you all, now I understand why it is on only one network :D

**Reply by AJenbo** [reply]
>  In NBlood we don't have such discoverability, so we would need to implement something anyways.

We re-implemented all of the network functionality so it's not related to what was originally there in D1. It's actually pretty simple to do so.

A master server can have some benefits. But since the game is inherently p2p anyway it doesn't actually fit the design very well.

Besides libzt (which is mostly about connecting and communicating on the network) there is also the ZeroTier API which lets you do managment of your network and you could use docker based ZeroTier controlers to spin up new networks. But at that point you are moving out of the free tier and require a server infrastructure to run things on. If who ever runs that system decides to move on the whole multiplayer system goes with them. With the current solution we only have a dependency on ZeroTier which is unavoidable (for the ZeroTier protocol, there would still be direct TCP as an option).

Comparing that to libsodium which only requirement is a linker and have the added benefit of working on all supported network protocols it becomes hard to justify the extra work needed for going the other route. I suppose libsodium does add a bit of overhead on the CPU, but we have it running on 3DS and we are also doing dumb stuff like rendering on the CPU at the same time.
